src/main/java/org/mule/extension/file/internal/source/DirectoryListener.java

```diff
@@ -331,4 +311,100 @@ private LocalDateTime getWatermarkTimestamp(LocalFileAttributes attributes) {
+        pendingFilesByTimeCheck =
+            processFilesAndCalculateFilesPendingFromProcessDueSizeCheck(pendingFilesByTimeCheck, timeBetweenSizeCheckInMillis,
+                                                                        pollContext);
+        retries++;
+      }
+    }
+
+  }
+
+  private Map<String, Result<InputStream, LocalFileAttributes>> processFilesAndCalculateFilesPendingFromProcessDueSizeCheck(final Map<String, Result<InputStream, LocalFileAttributes>> filesToProcess,
+                                                                                                                            final Long timeBetweenSizeCheckInMillis,
+                                                                                                                            final PollContext<InputStream, LocalFileAttributes> pollContext) {
+    try {
+      Map<String, Result<InputStream, LocalFileAttributes>> pendingFilesByTimeCheck = new HashMap<>();
+
+      if (filesToProcess.isEmpty()) {
+        return pendingFilesByTimeCheck;// no files to process
+      }
+
+      if (isNotNull(timeBetweenSizeCheckInMillis) && timeBetweenSizeCheckInMillis > 0) {
+        sleep(timeBetweenSizeCheckInMillis);
+      }
+
+      List<Result<InputStream, LocalFileAttributes>> currentList =
+          fileSystem.list(config, directoryPath.toString(), recursive, matcher, timeBetweenSizeCheckInMillis, null);
+      Map<String, Result<InputStream, LocalFileAttributes>> currentFilesMap = toMap(currentList);
+
+      if (currentList.isEmpty()) {
+        return pendingFilesByTimeCheck;// files already processed
+      }
+
+      Map<String, Result<InputStream, LocalFileAttributes>> filteredOldMap = filesToProcess.entrySet().stream()
+          .filter(entry -> currentFilesMap.containsKey(entry.getKey()))
+          .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+
+      processFiles(pollContext, pendingFilesByTimeCheck, currentFilesMap, filteredOldMap);
+      return pendingFilesByTimeCheck;
+    } catch (InterruptedException e) {
+      throw new MuleRuntimeException(createStaticMessage("Execution was interrupted while waiting to recheck file sizes"), e);
+    }
+  }
+
+  private void processFiles(PollContext<InputStream, LocalFileAttributes> pollContext,
+                            Map<String, Result<InputStream, LocalFileAttributes>> pendingFilesByTimeCheck,
+                            Map<String, Result<InputStream, LocalFileAttributes>> currentFilesMap,
+                            Map<String, Result<InputStream, LocalFileAttributes>> filteredOldMap) {
+    boolean sourceIsStopping = false;
+    for (final Map.Entry<String, Result<InputStream, LocalFileAttributes>> file : filteredOldMap.entrySet()) {
+      if (sourceIsStopping) {
+        closeResultQuietly(file.getValue());
+        continue;
+      }
+
+      Result<InputStream, LocalFileAttributes> oldInputStreamLocalFileAttributesResult = filteredOldMap.get(file.getKey());
```

Comment:
The variable file has the result so you can get there with file.getValue(), you don't have to get the value from the hashmap again.